tower/src/util/combinators/map_err.rs

```diff
@@ -0,0 +1,70 @@
+use futures_util::{future::MapErr as MapErrFut, TryFutureExt};
+use std::marker::PhantomData;
+use std::task::{Context, Poll};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Service returned by the [`map_err`] combinator.
+///
+/// [`map_err`]: crate::util::ServiceExt::map_err
+#[derive(Debug)]
+pub struct MapErr<S, F> {
+    inner: S,
+    f: F,
+}
+
+impl<S, F> MapErr<S, F> {
+    /// Creates a new [`MapErr`] service.
+    pub fn new(inner: S, f: F) -> Self {
+        MapErr { f, inner }
+    }
+}
+
+impl<S, F, Request, Error> Service<Request> for MapErr<S, F>
+where
+    S: Service<Request>,
+    F: FnOnce(S::Error) -> Error + Clone,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = MapErrFut<S::Future, F>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(self.f.clone())
+    }
+
+    fn call(&mut self, request: Request) -> Self::Future {
+        self.inner.call(request).map_err(self.f.clone())
+    }
+}
+
+/// A [`Layer`] that produces a [`MapErr`] service.
+///
+/// [`Layer`]: tower_layer::Layer
+#[derive(Debug)]
+pub struct MapErrLayer<F, Request, Error> {
+    f: F,
+    _p: PhantomData<fn(Request, Error)>,
+}
+
+impl<F, Request, Error> MapErrLayer<F, Request, Error> {
+    /// Creates a new [`MapErrLayer`].
+    pub fn new(f: F) -> Self {
+        MapErrLayer { f, _p: PhantomData }
+    }
+}
+
+impl<S, F, Request, Error> Layer<S> for MapErrLayer<F, Request, Error>
+where
+    S: Service<Request>,
+    F: FnOnce(S::Error) -> Error + Clone,
```

Comment:
This is similar to what I did here https://github.com/tower-rs/tower/blob/master/tower/src/util/map/response.rs#L83

The reason I went with just `F` clone here is to remove the need to have the generic for the Request/Error type in the layer implementation. This removes the need for the phantom data. I would say we should probably follow how that implementation does it.